File: src/board.rs
//! Board representation, pieces, and coordinate helpers.
//!
//! Coordinate system: 9 files (x: 0..=8) by 10 ranks (y: 0..=9).
//! `index = rank * 9 + file`. Rank 0 is the top (Black's home rank);
//! rank 9 is the bottom (Red's home rank). Red moves first and sits at the
//! bottom, so Red soldiers advance toward smaller ranks.

use serde::{Deserialize, Serialize};

pub const FILES: usize = 9;
pub const RANKS: usize = 10;
pub const CELLS: usize = FILES * RANKS;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum Color {
    Red,
    Black,
}

impl Color {
    pub fn opposite(self) -> Color {
        match self {
            Color::Red => Color::Black,
            Color::Black => Color::Red,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum PieceKind {
    General,
    Advisor,
    Elephant,
    Horse,
    Chariot,
    Cannon,
    Soldier,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub struct Piece {
    pub kind: PieceKind,
    pub color: Color,
}

impl Piece {
    /// Traditional glyph used to draw the piece.
    pub fn glyph(self) -> &'static str {
        use Color::*;
        use PieceKind::*;
        match (self.kind, self.color) {
            (General, Red) => "帥",
            (General, Black) => "將",
            (Advisor, Red) => "仕",
            (Advisor, Black) => "士",
            (Elephant, Red) => "相",
            (Elephant, Black) => "象",
            (Horse, _) => "馬",
            (Chariot, _) => "車",
            (Cannon, Red) => "炮",
            (Cannon, Black) => "砲",
            (Soldier, Red) => "兵",
            (Soldier, Black) => "卒",
        }
    }

    /// Latin one-letter label, used when no CJK font is available so the
    /// board stays playable. Colour is conveyed by the disc/ink colour.
    pub fn ascii(self) -> &'static str {
        match self.kind {
            PieceKind::General => "K",
            PieceKind::Advisor => "A",
            PieceKind::Elephant => "E",
            PieceKind::Horse => "H",
            PieceKind::Chariot => "R",
            PieceKind::Cannon => "C",
            PieceKind::Soldier => "P",
        }
    }
}

#[inline]
pub fn idx(file: i32, rank: i32) -> usize {
    (rank as usize) * FILES + (file as usize)
}

#[inline]
pub fn file_of(i: usize) -> i32 {
    (i % FILES) as i32
}

#[inline]
pub fn rank_of(i: usize) -> i32 {
    (i / FILES) as i32
}

#[inline]
pub fn on_board(file: i32, rank: i32) -> bool {
    file >= 0 && file < FILES as i32 && rank >= 0 && rank < RANKS as i32
}

/// Is `(file, rank)` inside the 3x3 palace of `color`?
pub fn in_palace(color: Color, file: i32, rank: i32) -> bool {
    if !(3..=5).contains(&file) {
        return false;
    }
    match color {
        Color::Black => (0..=2).contains(&rank),
        Color::Red => (7..=9).contains(&rank),
    }
}

/// Has the piece at `rank` crossed the river, for the given color?
pub fn crossed_river(color: Color, rank: i32) -> bool {
    match color {
        Color::Red => rank <= 4,
        Color::Black => rank >= 5,
    }
}

/// Is `rank` on `color`'s own half (used to keep elephants from crossing)?
pub fn own_half(color: Color, rank: i32) -> bool {
    match color {
        Color::Red => rank >= 5,
        Color::Black => rank <= 4,
    }
}

/// Forward rank direction for a soldier of `color` (Red advances up = -1).
pub fn forward(color: Color) -> i32 {
    match color {
        Color::Red => -1,
        Color::Black => 1,
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub struct Move {
    pub from: usize,
    pub to: usize,
}

#[derive(Debug, Clone, Copy)]
pub struct Board {
    pub cells: [Option<Piece>; CELLS],
}

impl Board {
    pub fn empty() -> Self {
        Board {
            cells: [None; CELLS],
        }
    }

    /// Standard opening position.
    pub fn initial() -> Self {
        let mut b = Board::empty();
        let back = [
            PieceKind::Chariot,
            PieceKind::Horse,
            PieceKind::Elephant,
            PieceKind::Advisor,
            PieceKind::General,
            PieceKind::Advisor,
            PieceKind::Elephant,
            PieceKind::Horse,
            PieceKind::Chariot,
        ];

        // Black at the top (ranks 0..=3).
        for (f, &kind) in back.iter().enumerate() {
            b.cells[idx(f as i32, 0)] = Some(Piece {
                kind,
                color: Color::Black,
            });
        }
        b.cells[idx(1, 2)] = Some(Piece { kind: PieceKind::Cannon, color: Color::Black });
        b.cells[idx(7, 2)] = Some(Piece { kind: PieceKind::Cannon, color: Color::Black });
        for f in (0..9).step_by(2) {
            b.cells[idx(f, 3)] = Some(Piece { kind: PieceKind::Soldier, color: Color::Black });
        }

        // Red at the bottom (ranks 6..=9).
        for (f, &kind) in back.iter().enumerate() {
            b.cells[idx(f as i32, 9)] = Some(Piece {
                kind,
                color: Color::Red,
            });
        }
        b.cells[idx(1, 7)] = Some(Piece { kind: PieceKind::Cannon, color: Color::Red });
        b.cells[idx(7, 7)] = Some(Piece { kind: PieceKind::Cannon, color: Color::Red });
        for f in (0..9).step_by(2) {
            b.cells[idx(f, 6)] = Some(Piece { kind: PieceKind::Soldier, color: Color::Red });
        }

        b
    }

    #[inline]
    pub fn get(&self, i: usize) -> Option<Piece> {
        self.cells[i]
    }

    /// Apply a move, returning the captured piece (if any) for later undo.
    pub fn make(&mut self, mv: Move) -> Option<Piece> {
        let captured = self.cells[mv.to];
        self.cells[mv.to] = self.cells[mv.from];
        self.cells[mv.from] = None;
        captured
    }

    /// Reverse `make`.
    pub fn unmake(&mut self, mv: Move, captured: Option<Piece>) {
        self.cells[mv.from] = self.cells[mv.to];
        self.cells[mv.to] = captured;
    }

    pub fn find_general(&self, color: Color) -> Option<usize> {
        self.cells.iter().position(|c| {
            matches!(c, Some(p) if p.kind == PieceKind::General && p.color == color)
        })
    }
}

File: alphazero/xiangqi/board.py
"""Board representation and coordinate helpers.

The board is a flat ``numpy.int8`` array of length 90. Each cell holds a
piece *code*:

    0                      empty
    1 + kind*2 + color     occupied  (kind 0..6, color 0=Red 1=Black)

This mirrors the encoding used by ``game.rs::position_hash`` so a board can
be hashed by its raw bytes for repetition detection.
"""

from __future__ import annotations

import numpy as np

# Colors.
RED = 0
BLACK = 1

# Piece kinds (same order as the Rust ``PieceKind`` enum).
GENERAL = 0
ADVISOR = 1
ELEPHANT = 2
HORSE = 3
CHARIOT = 4
CANNON = 5
SOLDIER = 6

FILES = 9
RANKS = 10
CELLS = FILES * RANKS  # 90


def opposite(color: int) -> int:
    return color ^ 1


def idx(file: int, rank: int) -> int:
    return rank * FILES + file


def file_of(i: int) -> int:
    return i % FILES


def rank_of(i: int) -> int:
    return i // FILES


def on_board(file: int, rank: int) -> bool:
    return 0 <= file < FILES and 0 <= rank < RANKS


def in_palace(color: int, file: int, rank: int) -> bool:
    if not (3 <= file <= 5):
        return False
    if color == BLACK:
        return 0 <= rank <= 2
    return 7 <= rank <= 9  # RED


def crossed_river(color: int, rank: int) -> bool:
    return rank <= 4 if color == RED else rank >= 5


def own_half(color: int, rank: int) -> bool:
    return rank >= 5 if color == RED else rank <= 4


def forward(color: int) -> int:
    """Forward rank direction for a soldier (Red advances toward rank 0)."""
    return -1 if color == RED else 1


def encode_piece(kind: int, color: int) -> int:
    return 1 + kind * 2 + color


def decode_piece(code: int):
    """Return ``(kind, color)`` for a non-empty code, else ``None``."""
    if code == 0:
        return None
    c = code - 1
    return (c >> 1, c & 1)


def piece_color(code: int) -> int:
    return (code - 1) & 1


def piece_kind(code: int) -> int:
    return (code - 1) >> 1


def empty_board() -> np.ndarray:
    return np.zeros(CELLS, dtype=np.int8)


def initial_board() -> np.ndarray:
    """Standard opening position (Black on ranks 0..3, Red on ranks 6..9)."""
    b = empty_board()
    back = [CHARIOT, HORSE, ELEPHANT, ADVISOR, GENERAL,
            ADVISOR, ELEPHANT, HORSE, CHARIOT]

    for f, kind in enumerate(back):
        b[idx(f, 0)] = encode_piece(kind, BLACK)
    b[idx(1, 2)] = encode_piece(CANNON, BLACK)
    b[idx(7, 2)] = encode_piece(CANNON, BLACK)
    for f in range(0, 9, 2):
        b[idx(f, 3)] = encode_piece(SOLDIER, BLACK)

    for f, kind in enumerate(back):
        b[idx(f, 9)] = encode_piece(kind, RED)
    b[idx(1, 7)] = encode_piece(CANNON, RED)
    b[idx(7, 7)] = encode_piece(CANNON, RED)
    for f in range(0, 9, 2):
        b[idx(f, 6)] = encode_piece(SOLDIER, RED)

    return b


def find_general(board: np.ndarray, color: int) -> int:
    """Cell index of ``color``'s general, or -1 if it is gone."""
    target = encode_piece(GENERAL, color)
    hits = np.nonzero(board == target)[0]
    return int(hits[0]) if hits.size else -1
